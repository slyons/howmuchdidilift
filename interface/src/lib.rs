use serde::{Deserialize, Serialize};
use strum::{EnumIter, EnumString};
use chrono::{DateTime, Utc};

#[derive(Debug, Deserialize, Serialize)]
pub struct LoginParams {
    pub email: String,
    pub password: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct LoginResponse {
    pub token: String,
    pub pid: String,
    pub name: String,
    pub is_verified: bool,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct CurrentResponse {
    pub pid: String,
    pub name: String,
    pub email: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct RegisterParams {
    pub email: String,
    pub password: String,
    pub password_confirm: String,
    pub name: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct VerifyParams {
    pub token: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct ForgotParams {
    pub email: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct ResetParams {
    pub token: String,
    pub password: String,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct MeasureCreate {
    pub name: String,
    pub name_plural: String,
    pub grams: f64,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, Default)]
pub struct Measure {
    pub id: i32,
    pub name: String,
    pub name_plural: String,
    pub grams: f64,
}

#[derive(Debug, Clone, Deserialize, Serialize, EnumIter, EnumString)]
#[serde(rename_all = "lowercase")]
pub enum InputWeightType {
    Lbs,
    Kgs,
}

#[derive(Debug, Deserialize, Serialize)]
pub struct RandomWeightRequest {
    pub input_amt: f64,
    pub input_type: InputWeightType,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct RandomWeightResponse {
    pub when: DateTime<Utc>,
    pub input_amt: f64,
    pub input_type: InputWeightType,
    pub output_weight: String,
}


