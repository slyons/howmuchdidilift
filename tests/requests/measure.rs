use axum::Json;
use insta::{assert_debug_snapshot, with_settings};
use interface::{InputWeightType, MeasureCreate, RandomWeightRequest, RandomWeightResponse};
use liftcalc::app::App;
use loco_rs::testing;
use serial_test::serial;

use crate::requests::prepare_data;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("measure_request");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_create() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        //testing::seed::<App>(&ctx.db).await.unwrap();
        let user = prepare_data::init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = prepare_data::auth_header(&user.token);

        let create_request = MeasureCreate {
            name: "cheeseburger".to_string(),
            name_plural: "cheeseburgers".to_string(),
            grams: 500.0,
        };
        let measures = request
            .post("/api/measures")
            .add_header(auth_key, auth_value)
            .json(&create_request)
            .await;

        with_settings!({
            filters => {
                 let mut combined_filters = testing::CLEANUP_DATE.to_vec();
                    combined_filters.extend(vec![(r#"\"id\\":\d+"#, r#""id\":ID"#)]);
                    combined_filters
            }
        }, {
            assert_debug_snapshot!(
                (measures.status_code(), measures.text())
            );
        });
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_convert_lbs() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        let user = prepare_data::init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = prepare_data::auth_header(&user.token);

        let create_request = MeasureCreate {
            name: "gramburger".to_string(),
            name_plural: "gramburgers".to_string(),
            grams: 500.0,
        };
        let measures = request
            .post("/api/measures")
            .add_header(auth_key, auth_value)
            .json(&create_request)
            .await;
        measures.assert_status_ok();

        let convert_request = RandomWeightRequest {
            input_amt: 100.0,
            input_type: InputWeightType::Lbs,
        };
        let random_weight = request
            .post("/api/measures/convert")
            .json(&convert_request)
            .await;
        random_weight.assert_status_ok();

        let random_weight: RandomWeightResponse = random_weight.json();
        assert_debug_snapshot!(random_weight);
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_convert_kgs() {
    configure_insta!();

    testing::request::<App, _, _>(|request, ctx| async move {
        let user = prepare_data::init_user_login(&request, &ctx).await;
        let (auth_key, auth_value) = prepare_data::auth_header(&user.token);

        let create_request = MeasureCreate {
            name: "gramburger".to_string(),
            name_plural: "gramburgers".to_string(),
            grams: 1.0,
        };
        let measures = request
            .post("/api/measures")
            .add_header(auth_key.clone(), auth_value.clone())
            .json(&create_request)
            .await;
        measures.assert_status_ok();

        let convert_request = RandomWeightRequest {
            input_amt: 1.0,
            input_type: InputWeightType::Kgs,
        };
        let random_weight = request
            .post("/api/measures/convert")
            .add_header(auth_key, auth_value)
            .json(&convert_request)
            .await;
        random_weight.assert_status_ok();

        let random_weight: RandomWeightResponse = random_weight.json();
        assert_debug_snapshot!(random_weight);
    })
    .await;
}
