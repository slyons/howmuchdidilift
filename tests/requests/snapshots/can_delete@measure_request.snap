---
source: tests/requests/measure.rs
expression: "(all_measures, measures)"
---
(
    [
        Measure {
            id: 1,
            name: "grams",
            grams: 1.0,
        },
    ],
    [],
)
